### PR TITLE
Dev/257: Query the exam's transcript

### PR DESCRIPTION
**Pull request by Johnny850807**
Dev/257: Query the exam's transcript (calculate the exam's score on request)

image

### CODE REVIEW

File: API/Submission-API/src/main/java/tw/waterball/judgegirl/submissionapi/views/VerdictView.java [hunk 1]
```diff
@@ -46,29 +46,40 @@ public static VerdictView toViewModel(@Nullable Verdict verdict) {
...
+                return Verdict.compileError(verdictView.compileErrorMessage,
+                        verdictView.maxGrade, verdictView.issueTime);
+            }
+            Verdict verdict = new Verdict(
+                    mapToList(verdictView.getJudges(), JudgeView::toEntity),
+                    verdictView.getIssueTime());
+            verdict.setReport(verdictView.getReport().toEntity());
+            return verdict;
+        } catch (Exception err) {
+            return null;
```
Comment (Wally5077) [inline API/Submission-API/src/main/java/tw/waterball/judgegirl/submissionapi/views/VerdictView.java:82]: 這邊為什麼需要 try-catch?

File: Commons/src/main/java/tw/waterball/judgegirl/commons/utils/functional/GetById.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package tw.waterball.judgegirl.commons.utils.functional;
+
+/**
+ * @author - johnny850807@gmail.com (Waterball)
+ */
+@FunctionalInterface
+public interface GetById<ID, T> {
```
Comment (Wally5077) [inline Commons/src/main/java/tw/waterball/judgegirl/commons/utils/functional/GetById.java:7]: 這個是否為可用 Function apply 取代?

File: Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CalculateExamScoreUseCase.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+        var examinees = examinees(exam);
+        List<ExamineeRecord> examineeRecords = examineeRecords(exam, records, examinees);
+        double averageScore = average(examineeRecords, ExamineeRecord::getTotalScore);
+        present(presenter, exam, examineeRecords, averageScore);
+    }
+
+    private List<Record> filterRecords(List<Record> records, GetById<Integer, Student> examinees) {
+        return filterToList(records, record -> examinees.get(record.getStudentId()) != null);
+    }
```
Comment (Wally5077) [inline Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CalculateExamScoreUseCase.java:43]: 這個方法似乎沒有用到，可以考慮刪掉

File: Commons/src/main/java/tw/waterball/judgegirl/commons/utils/StreamUtils.java [hunk 4]
```diff
@@ -84,9 +81,28 @@
         return collection.stream().collect(Collectors.toMap(keyMapper, valueMapper));
     }
 
+    public static <T, L, R> List<T> zipOneToMany(L left, List<R> right,
+                                                 BiFunction<L, R, T> zipAndMap) {
+        return right.stream()
```
Comment (Wally5077) [inline Commons/src/main/java/tw/waterball/judgegirl/commons/utils/StreamUtils.java:86]: return mapToList(right, r -> zipAndMap.apply(left, r));

File: Commons/src/main/java/tw/waterball/judgegirl/commons/utils/StreamUtils.java [hunk 5]
```diff
@@ -84,9 +81,28 @@
...
+    }
+
+    public static <T, L, R> List<T> zipToList(Map<L, R> map,
+                                              BiFunction<L, R, T> zipAndMap) {
+        return zipToList(map.entrySet(), zipAndMap);
+    }
+
+    public static <T, L, R> List<T> zipToList(Collection<Map.Entry<L, R>> entrySet,
+                                              BiFunction<L, R, T> zipAndMap) {
+        return entrySet.stream()
```
Comment (Wally5077) [inline Commons/src/main/java/tw/waterball/judgegirl/commons/utils/StreamUtils.java:98]: return mapToList(entrySet, entry -> zipAndMap.apply(entry.getKey(), entry.getValue()));

File: Domain/Primitives/src/main/java/tw/waterball/judgegirl/primitives/stubs/SubmissionStubBuilder.java [hunk 6]
```diff
@@ -112,22 +119,28 @@ private void lazyInitializeVerdictIfNull() {
...
-                    statuses.get(i) == AC ? gradePerAc : 0));
-        }
-        Submission submission = new Submission(UUID.randomUUID().toString(), studentId, problem.getId(),
-                Language.C.toString(), "fileId");
+        List<JudgeStatus> statuses = generate(howManyACs, AC);
+        statuses.addAll(generate(problem.numOfTestcases() - howManyACs,
+                i -> NORMAL_STATUSES_NO_CE[random.nextInt(NORMAL_STATUSES_NO_CE.length)]));
+
+        List<Judge> judges = generate(problem.numOfTestcases(),
+                i -> new Judge(problem.getTestcases().get(i),
```
Comment (Wally5077) [inline Domain/Primitives/src/main/java/tw/waterball/judgegirl/primitives/stubs/SubmissionStubBuilder.java:137]: problem.getTestcases().get(i) ->problem.getTestcase(i)

File: Judger/src/main/java/tw/waterball/judgegirl/judger/tests/AbstractJudgerTest.java [hunk 7]
```diff
@@ -177,7 +177,7 @@ private void verifyTLEPublished() {
     private void verifyWAPublished() {
         VerdictIssuedEvent event = captureVerdictIssuedEvent();
         Verdict verdict = event.getVerdict();
-        for (int i = 0; i < problem.getTestcases().size(); i++) {
+        for (int i = 0; i < problem.numOfTestcases(); i++) {
             Judge judge = verdict.getJudges().get(i);
             Testcase testCase = problem.getTestcases().get(i);
```
Comment (Wally5077) [inline Judger/src/main/java/tw/waterball/judgegirl/judger/tests/AbstractJudgerTest.java:182]: ditto

File: Judger/src/main/java/tw/waterball/judgegirl/judger/tests/AbstractJudgerTest.java [hunk 8]
```diff
@@ -129,7 +129,7 @@ void judge_WA() throws IOException {
     private void verifyACPublished() {
         VerdictIssuedEvent event = captureVerdictIssuedEvent();
         Verdict verdict = event.getVerdict();
-        for (int i = 0; i < problem.getTestcases().size(); i++) {
+        for (int i = 0; i < problem.numOfTestcases(); i++) {
             Judge judge = verdict.getJudges().get(i);
             Testcase testCase = problem.getTestcases().get(i);
```
Comment (Wally5077) [inline Judger/src/main/java/tw/waterball/judgegirl/judger/tests/AbstractJudgerTest.java:134]: ditto (SubmissionStubBuilder)

File: Judger/src/main/java/tw/waterball/judgegirl/judger/tests/AbstractJudgerTest.java [hunk 9]
```diff
@@ -159,7 +159,7 @@ private void verifyCEPublished() {
     private void verifyTLEPublished() {
         VerdictIssuedEvent event = captureVerdictIssuedEvent();
         Verdict verdict = event.getVerdict();
-        for (int i = 0; i < problem.getTestcases().size(); i++) {
+        for (int i = 0; i < problem.numOfTestcases(); i++) {
             Judge judge = verdict.getJudges().get(i);
             Testcase testCase = problem.getTestcases().get(i);
```
Comment (Wally5077) [inline Judger/src/main/java/tw/waterball/judgegirl/judger/tests/AbstractJudgerTest.java:164]: ditto

File: Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/ExamControllerTest.java [hunk 10]
```diff
@@ -586,6 +577,61 @@ void GivenOneExamCreated_WhenDeleteTheExam_ShouldSucceed() throws Exception {
...
+    public static void initTest() throws SQLException {
+        Server.createWebServer("-web", "-webAllowOthers", "-webPort", "8082")
+                .start();
+    }
+
+    // TODO: drunk code, need to be improved
+    @Test
+    void testProduceExamTranscript() throws Exception {
+        Integer[] studentIds = {STUDENT_A_ID, STUDENT_B_ID, STUDENT_C_ID, STUDENT_D_ID};
+        var exam = createExamAndGet(new Date(), oneSecondAfter(), "Exam");
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/ExamControllerTest.java:590]: var exam = createExamAndGet(now(), oneSecondAfter(), "Exam"); 

備註: (DateProvider.now())

File: Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/ExamControllerTest.java [hunk 11]
```diff
@@ -586,6 +577,61 @@ void GivenOneExamCreated_WhenDeleteTheExam_ShouldSucceed() throws Exception {
...
+        int scoreB1 = q1.calculateScore(sB1), scoreB2 = q2.calculateScore(sB2);
+        int scoreC1 = q1.calculateScore(sC1), scoreC2 = q2.calculateScore(sC2);
+
+        assertEquals(q1.getScore() + q2.getScore(), transcript.getMaxScore());
+        assertEquals(scoreA1 + scoreA2, examineeRecordA.totalScore);
+        assertEqualsIgnoreOrder(asList(scoreA1, scoreA2), examineeRecordA.questionScores);
+        assertEquals(scoreB1 + scoreB2, examineeRecordB.totalScore);
+        assertEqualsIgnoreOrder(asList(scoreB1, scoreB2), examineeRecordB.questionScores);
+        assertEquals(scoreC1 + scoreC2, examineeRecordC.totalScore);
+        assertEqualsIgnoreOrder(asList(scoreC1, scoreC2), examineeRecordC.questionScores);
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/ExamControllerTest.java:617]: 請抽出個驗證方法

File: Spring-Boot/Spring-Boot-Commons/src/test/java/tw/waterball/judgegirl/springboot/configs/JacksonConfigTest.java [hunk 12]
```diff
@@ -56,14 +56,24 @@ void testDeserializingReportView() throws JsonProcessingException {
...
-    void testDeserializingVerdictIssuedEvent() throws JsonProcessingException {
-        Verdict verdict = new Verdict(
+    void testWAVerdictIssuedEvent() throws JsonProcessingException {
+        testVerdictIssuedEventUsingObjectMapper(new Verdict(
                 asList(new Judge("T", JudgeStatus.AC, new ProgramProfile(10, 10, ""),
                                 new Grade(50, 50)),
                         new Judge("T", JudgeStatus.WA, new ProgramProfile(10, 100, ""),
                                 new Grade(0, 50))
-                ), new Date());
+                ), new Date()));
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Commons/src/test/java/tw/waterball/judgegirl/springboot/configs/JacksonConfigTest.java:73]: ditto (new Date -> now())

File: API/Submission-API/src/main/java/tw/waterball/judgegirl/submissionapi/views/VerdictView.java (hunk 1 above)
Comment (Johnny850807) [reply API/Submission-API/src/main/java/tw/waterball/judgegirl/submissionapi/views/VerdictView.java:82]: Oh fffff, I forgot to delete it, thank you!!!!!

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

File: Commons/src/main/java/tw/waterball/judgegirl/commons/utils/functional/GetById.java (hunk 2 above)
Comment (Johnny850807) [reply Commons/src/main/java/tw/waterball/judgegirl/commons/utils/functional/GetById.java:7]: the term `apply` is ugly.

I prefer using this interface.

File: Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CalculateExamScoreUseCase.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        presenter.showExam(exam);
+        examineeRecords.forEach(presenter::addRecord);
+        presenter.showStatistics(averageScore, exam.getMaxScore());
+    }
+
+    public interface Presenter {
+
+        void showExam(Exam exam);
+
+        void addRecord(ExamineeRecord examineeRecord);
```
Comment (Johnny850807) [inline Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CalculateExamScoreUseCase.java:72]: @Wally5077 Note, this is the new naming convention (`addXXX`), when we are going to present multiple elements.

File: Spring-Boot/Spring-Boot-Commons/src/test/java/tw/waterball/judgegirl/springboot/configs/JacksonConfigTest.java (hunk 12 above)
Comment (Johnny850807) [reply Spring-Boot/Spring-Boot-Commons/src/test/java/tw/waterball/judgegirl/springboot/configs/JacksonConfigTest.java:73]: great catch!!!!

**Pull request merged**